### PR TITLE
Extensions UBOs

### PR DESCRIPTION
**Pull request by felixpalmer**
For https://github.com/visgl/deck.gl/issues/8997

#### Change List
- Port `BrushingExtension`
- Port `ClipExtension`
- Port `CollisionFilterExtension`
- Port `FillStyleExtension`
- Port `PathStyleExtension`
- Port `TerrainExtension`

### CODE REVIEW

File: modules/extensions/src/clip/clip-extension.ts [hunk 1]
```diff
@@ -77,9 +86,12 @@ in float clip_isVisible;
  * The fragment-shader version clips pixels at the bounds
  * e.g. PolygonLayer - show the part of the polygon that intersect with the bounds
  */
-const shaderModuleFs: ShaderModule = {
-  name: 'clip-fs',
-  fs: shaderFunction
+const shaderModuleFs: ShaderModule<ClipModuleSettings> = {
+  name: 'clip',
```
Comment (felixpalmer) [inline modules/extensions/src/clip/clip-extension.ts:90]: Any risk sharing the name between the modules?

File: modules/extensions/src/brushing/brushing-extension.ts [hunk 2]
```diff
@@ -101,4 +101,17 @@ export default class BrushingExtension extends LayerExtension {
       });
     }
   }
+
+  draw(this: Layer<BrushingExtensionProps>, params: any, extension: this) {
+    const {viewport, mousePosition} = params.context;
+    const {brushingEnabled, brushingRadius, brushingTarget} = params.moduleParameters;
+    const brushingProps: BrushingModuleSettings = {
```
Comment (ibgreen) [inline modules/extensions/src/brushing/brushing-extension.ts:108]: Nit: variable name should follow type. We already have some confusion between "props" and "settings", this makes it worse
```suggestion
// either
    const brushingProps: BrushingModuleProps = {
// or
    const brushingSettings: BrushingModuleSettings = {
```

File: modules/extensions/src/brushing/brushing-extension.ts [hunk 3]
```diff
@@ -101,4 +101,17 @@ export default class BrushingExtension extends LayerExtension {
...
+      viewport,
+      mousePosition,
+      brushingEnabled,
+      brushingRadius,
+      brushingTarget
+    };
+    this.setShaderModuleProps({brushing: brushingProps});
```
Comment (ibgreen) [inline modules/extensions/src/brushing/brushing-extension.ts:115]: Again, the function is called `set...Props` but it takes `...Settings`

In luma I have been trying to standardize on `...Props` for shader modules.

File: modules/extensions/src/clip/clip-extension.ts (hunk 1 above)
Comment (ibgreen) [reply modules/extensions/src/clip/clip-extension.ts:90]: > Any risk sharing the name between the modules?

I am not quite sure what the context is but I would avoid using same name for two modules if that is the question.

**Review by ibgreen — APPROVED** [review]
LGTM.

File: modules/extensions/src/clip/clip-extension.ts (hunk 1 above)
Comment (felixpalmer) [reply modules/extensions/src/clip/clip-extension.ts:90]: [Luma assumes](https://github.com/visgl/luma.gl/blob/4e28a224d0f2f7990ee107bfa52044e09106a68b/modules/engine/src/model/model.ts#L519) that the uniform block will be called `${moduleName}Uniforms` - which for starters means we cannot use `clip-fs`. 

It also means that if we cannot share the same uniformBlock across multiple modules. Perhaps we should add an optional `ShaderModule.uniformBufferName` in luma?

File: modules/extensions/src/brushing/brushing-extension.ts (hunk 3 above)
Comment (felixpalmer) [reply modules/extensions/src/brushing/brushing-extension.ts:115]: Renamed across all extensions

**Pull request merged**